### PR TITLE
Properties endpoints

### PR DESCRIPTION
**Pull request by andrewgwoodruff**
This adds `put` and `get` endpoints for `/properties`, per https://github.com/open-realms/express-minecraft/issues/3.

### CODE REVIEW

File: minecraft.js [hunk 1]
```diff
@@ -66,6 +66,25 @@ app.post('/command', (req, res) => {
   }, 250);
 });
 
+app.get('/properties', (req, res) => {
+  const text = fs.readFileSync('./server.properties').toString('utf-8');
```
Comment (johnbwoodruff) [inline minecraft.js]: You already mentioned this, but yeah don't use the sync version.

Comment (johnbwoodruff) [reply minecraft.js]: You can make a promise version of the async version so you can use async/await

File: minecraft.js [hunk 2]
```diff
@@ -96,12 +115,33 @@ function shutdownMinecraft() {
...
-    startMinecraft();
   }
+  startMinecraft();
 }
 
 function getMinecraftStatus() {
   return serverStatus;
 }
+
+function fileToJSON(input) {
```
Comment (johnbwoodruff) [inline minecraft.js]:
```suggestion
function parseProperties(input) {
```

File: minecraft.js [hunk 3]
```diff
@@ -96,12 +115,33 @@ function shutdownMinecraft() {
...
+  var output = {};
+
+  input.split('\n').forEach(function (line) {
+    if (line[0] === '#') {
```
Comment (johnbwoodruff) [inline minecraft.js:132]: You should probably also handle empty lines. So if the line is an empty string, skip as well.

File: minecraft.js [hunk 4]
```diff
@@ -96,12 +115,33 @@ function shutdownMinecraft() {
...
+      return;
+    }
+
+    let parts = line.split('=');
+
+    if (parts[0] === '') {
```
Comment (johnbwoodruff) [inline minecraft.js:139]: What usecase does this handle?

File: minecraft.js [hunk 5]
```diff
@@ -96,12 +115,33 @@ function shutdownMinecraft() {
...
 }
 
 function getMinecraftStatus() {
   return serverStatus;
 }
+
+function fileToJSON(input) {
+  var output = {};
+
+  input.split('\n').forEach(function (line) {
```
Comment (johnbwoodruff) [inline minecraft.js]: I would prefer using an arrow function

File: minecraft.js [hunk 6]
```diff
@@ -66,6 +66,25 @@ app.post('/command', (req, res) => {
...
+app.get('/properties', (req, res) => {
+  const text = fs.readFileSync('./server.properties').toString('utf-8');
+  const result = fileToJSON(text);
+  res.send(result);
+});
+
+app.put('/properties', (req, res) => {
+  const properties = req.body;
+  let string = '#Minecraft server properties\n#(last boot timestamp)\n';
+  for (const property in properties) {
```
Comment (johnbwoodruff) [inline minecraft.js]: When using a `for...in` type loop typically I prefer to do `for (const key in properties)`

Comment (johnbwoodruff) [reply minecraft.js]: Makes it more clear you're referencing the key rather than the actual property.

File: minecraft.js [hunk 7]
```diff
@@ -66,6 +66,25 @@ app.post('/command', (req, res) => {
...
+    const sub_string = `${property}=${properties[property]}\n`;
+    string = string.concat(sub_string);
+  }
+  fs.writeFile('./server.properties', string, function (err) {
```
Comment (johnbwoodruff) [inline minecraft.js]: You should use the async version of this as well so you can await it and then return an error if the save fails.

File: minecraft.js [hunk 8]
```diff
@@ -66,6 +66,25 @@ app.post('/command', (req, res) => {
...
+    if (err) return console.log(err);
+  });
+  res.send({ status: 'server.properties successfully modified!' });
```
Comment (johnbwoodruff) [inline minecraft.js]: I'd return a status 204 no content and not return any body.

**Review by johnbwoodruff — CHANGES_REQUESTED** [review]

File: minecraft.js (hunk 4 above)
Comment (andrewgwoodruff) [reply minecraft.js:139]: It handles empty lines as well as the end of the file.

File: minecraft.js (hunk 3 above)
Comment (andrewgwoodruff) [reply minecraft.js:132]: That use case has already been added - it's the one you commented on below.

File: minecraft.js (hunk 6 above)
Comment (andrewgwoodruff) [reply minecraft.js]: Alrighty, that change has been made.

File: minecraft.js (hunk 5 above)
Comment (andrewgwoodruff) [reply minecraft.js]: K, I've made that change!

File: minecraft.js (hunk 8 above)
Comment (andrewgwoodruff) [reply minecraft.js]: Great idea! I've made that change.

File: minecraft.js (hunk 4 above)
Comment (johnbwoodruff) [reply minecraft.js:139]: Oh awesome

File: minecraft.js (hunk 1 above)
Comment (andrewgwoodruff) [reply minecraft.js]: Alrighty, I've made this change!

File: minecraft.js (hunk 7 above)
Comment (andrewgwoodruff) [reply minecraft.js]: I've made this change here as well!

File: minecraft.js [hunk 9]
```diff
@@ -66,6 +66,25 @@ app.post('/command', (req, res) => {
...
+    res.send(result);
+  } catch (e) {
+    res.status(500).send();
+  }
+});
+
+app.put('/properties', async (req, res) => {
+  try {
+    const properties = req.body;
+    writeProperties('./server.properties', properties);
```
Comment (johnbwoodruff) [inline minecraft.js]: Make sure you `await` this

Comment (andrewgwoodruff) [reply minecraft.js]: Ah yes, thank you for catching that!! I've made that change.

**Review by johnbwoodruff — APPROVED** [review]

**Pull request merged**